### DISCUSSION TITLE
How does Keycloak SSO work?

### DISCUSSION THREAD

**Discussion by JangYouJung** [keycloak/keycloak#31428] [Q&A]
I'm having trouble implementing Keycloak SSO and need some help. Currently, I have two clients within one realm and I expect that if I log into one service, I'll be automatically logged into the other service as well by Keycloak's SSO.

However, when I log into one service, I'm not automatically logged into the other. What could be the problem?

Let me explain the configuration for each client:
![스크린샷(67)](https://github.com/user-attachments/assets/4a94921f-b8eb-4142-83b4-0e6429b47699)

### [1] Springboot Login (OpenId)
The first client is connected to a React app. I have configured it so that when accessing the React app, the Keycloak login screen appears, and a token is retrieved from Keycloak. This token is then used for API communication on the backend.
![스크린샷(63)](https://github.com/user-attachments/assets/ee191ca9-b125-474d-8a63-c0886004e26b)

### [2] AWS (SAML)
The second client is registered as an Identity Provider in AWS IAM. I'm implementing a feature where the React app retrieves a Quicksight dashboard from AWS. I want to use Keycloak's SSO functionality for this, so I've registered it with the Quicksight Admin role in the AWS IAM Provider.
![스크린샷(64)](https://github.com/user-attachments/assets/3ccaf799-9c36-4526-955a-5e9a3f2e577b)
![스크린샷(66)](https://github.com/user-attachments/assets/568c912f-791c-49bb-96dc-aa2d6b4a5650)

The issue I'm facing is:
After logging into the React app with Keycloak, when I look at the embedded Quicksight part, instead of being automatically logged in via SSO, it redirects me to the 'Quicksight login screen'.
![키클록 로그인 에러](https://github.com/user-attachments/assets/fb9e893a-9a2f-4ab6-89ed-b84124784a0e)

So I want to know what I should do to **set up the SSO between react client and aws quicksight client.**
I'll be waiting for you response

**Comment by darius-m** [comment]
I don't think that any SSO service would be able to log you in automatically into a different service, at least not by default. The purpose of the SSO service in case of user authentication is twofold: it can enforce authentication policies (e.g., use of strong authentication) and keep a "central" session so users don't have to enter their credentials when logging in to every service.

Since services themselves do not communicate directly to share user sessions (mostly for security reasons, so sessions are not leaked), SSOs have been created. When you authenticate for the first time in a client, it redirect you to the SSO. The basic flow is the following:
1. the SSO checks if the user has a valid session (since it is the first authentication, the session does not exist, so it continues with the next two steps)
2. the service asks the user to log in, and the user enters their credentials to authenticate
3. if the information is correct, a session is created at SSO level
4. a response is sent to the client application;  the client can use the information in the response to create a session at application level (may require additional steps, for example if the code authentication flow in OIDC is used).

If the user authenticates a second time, the SSO checks the user session information (that would be using the alternative "cookie authentication" step in the authentication flow). Since the user already has a valid session, steps 2 and 3 skipped, and the response is sent to the client.

I'm not sure how authentication in the AWS service should happen in your case. If the application is completely separate, you'd probably have to find a mechanism to automatically redirect the user to login into the application. There are mechanisms for identity provider initiated logins, if that is allowed for AWS, which you could leverage.

However, if you do not need a client session for the other application (e.g., you only access a REST API), and you use the same protocol (I know how it woks for OIDC, not sure for SAML), you can probably define an extended audience for the tokens and use them as bearer tokens in your request to the third party application.
